llvm/test/CodeGen/RISCV/rvv/fixed-vectors-expandload-int.ll

```diff
@@ -250,50 +102,13 @@ declare <4 x i16> @llvm.masked.expandload.v4i16(ptr, <4 x i1>, <4 x i16>)
 define <4 x i16> @expandload_v4i16(ptr %base, <4 x i16> %src0, <4 x i1> %mask) {
 ; CHECK-LABEL: expandload_v4i16:
 ; CHECK:       # %bb.0:
-; CHECK-NEXT:    vsetivli zero, 1, e8, m1, ta, ma
-; CHECK-NEXT:    vmv.x.s a1, v0
-; CHECK-NEXT:    andi a2, a1, 1
-; CHECK-NEXT:    bnez a2, .LBB6_5
-; CHECK-NEXT:  # %bb.1: # %else
-; CHECK-NEXT:    andi a2, a1, 2
-; CHECK-NEXT:    bnez a2, .LBB6_6
-; CHECK-NEXT:  .LBB6_2: # %else2
-; CHECK-NEXT:    andi a2, a1, 4
-; CHECK-NEXT:    bnez a2, .LBB6_7
-; CHECK-NEXT:  .LBB6_3: # %else6
-; CHECK-NEXT:    andi a1, a1, 8
-; CHECK-NEXT:    bnez a1, .LBB6_8
-; CHECK-NEXT:  .LBB6_4: # %else10
-; CHECK-NEXT:    ret
-; CHECK-NEXT:  .LBB6_5: # %cond.load
-; CHECK-NEXT:    lh a2, 0(a0)
-; CHECK-NEXT:    vsetvli zero, zero, e16, m2, tu, ma
-; CHECK-NEXT:    vmv.s.x v8, a2
-; CHECK-NEXT:    addi a0, a0, 2
-; CHECK-NEXT:    andi a2, a1, 2
-; CHECK-NEXT:    beqz a2, .LBB6_2
-; CHECK-NEXT:  .LBB6_6: # %cond.load1
-; CHECK-NEXT:    lh a2, 0(a0)
-; CHECK-NEXT:    vsetvli zero, zero, e16, m2, ta, ma
-; CHECK-NEXT:    vmv.s.x v9, a2
-; CHECK-NEXT:    vsetivli zero, 2, e16, mf2, tu, ma
-; CHECK-NEXT:    vslideup.vi v8, v9, 1
-; CHECK-NEXT:    addi a0, a0, 2
-; CHECK-NEXT:    andi a2, a1, 4
-; CHECK-NEXT:    beqz a2, .LBB6_3
-; CHECK-NEXT:  .LBB6_7: # %cond.load5
-; CHECK-NEXT:    lh a2, 0(a0)
-; CHECK-NEXT:    vsetivli zero, 3, e16, mf2, tu, ma
-; CHECK-NEXT:    vmv.s.x v9, a2
-; CHECK-NEXT:    vslideup.vi v8, v9, 2
-; CHECK-NEXT:    addi a0, a0, 2
-; CHECK-NEXT:    andi a1, a1, 8
-; CHECK-NEXT:    beqz a1, .LBB6_4
-; CHECK-NEXT:  .LBB6_8: # %cond.load9
-; CHECK-NEXT:    lh a0, 0(a0)
-; CHECK-NEXT:    vsetivli zero, 4, e16, mf2, ta, ma
-; CHECK-NEXT:    vmv.s.x v9, a0
-; CHECK-NEXT:    vslideup.vi v8, v9, 3
+; CHECK-NEXT:    vsetivli zero, 4, e8, mf4, ta, ma
+; CHECK-NEXT:    vcpop.m a1, v0
+; CHECK-NEXT:    vsetvli zero, a1, e16, mf2, ta, ma
+; CHECK-NEXT:    vle16.v v9, (a0)
+; CHECK-NEXT:    vsetivli zero, 4, e16, mf2, ta, mu
+; CHECK-NEXT:    viota.m v10, v0
+; CHECK-NEXT:    vrgather.vv v8, v9, v10, v0.t
```

Comment:
Can this be done with a viota.m and an indexed load instead?

```asm
; v0 = 0 0 0 1 0 1 0 1
viota.m v9, v0 ; v9 = 3 3 3 2 2 1 1 0
vloxi64.v v8, a0, v9, v0.t ; v8 = v8.7, v8.6, v8.5, [a0+2], v8.3, [a0+1], v8.1, [a0+0]
```

Reply:
Good point! Done!
Both indexed load and vrgather are slow on current implementations, using indexed load has fewer instructions. And indexed load can be optimized in μ-arch level, but it may not be easy for vrgather.

Reply:
If anything, I'd imagine vrgather would be easier for hardware to optimize (though I am not a hardware designer), especially for smaller element types; I don't know of any architecture where indexed loads are anything better than one memory request per element (i.e. ~3 elements/cycle at best on already very high-end cores), but 16×i8 shuffles are usually able to run at at least 1/cycle (even down to ARM A53; and modern x86-64 cores (i.e. those with 3 loads/cycle) can even do 512-bit shuffles at 1/cycle). LMUL makes things somewhat more complicated, but I'd hope general-purpose cores would be designed in a way that doesn't make vrgather literally never usable.

If preferring indexed loads over load+vrgather was desired, it should also be done for, say, a reversing loop, but it's not (https://godbolt.org/z/5rPW6aoxq)

Reply:
A vrgather.vv is likely to be O(N^2) in terms of LMUL (since in the worst case scenario each register written to reads from LMUL registers), but vloxiN.v is O(N) in terms of VL. Or at least that's what we cost in RISCVTargetTransformInfo anyway, so we might as well be consistent there.

Reply:
@dzaima Thanks! Good information for other architectures!

For vrgather.vv, you can check https://camel-cdr.github.io/rvv-bench-results/index.html and see the costs of current implementations.
For indexed loads/stores, if they are consecutive (which is the case for this patch), these loads/stores can be combined in  μ-arch.

> If preferring indexed loads over load+vrgather was desired, it should also be done for, say, a reversing loop, but it's not (https://godbolt.org/z/5rPW6aoxq)

For reversed loop, we would prefer strided load with nagative stride, here we just haven't optimized for this case in vectorizer.

Reply:
I think the vector shuffle lowering runs into the same issue, it just bails if it can't promote the index to i16

https://github.com/llvm/llvm-project/blob/b1234ddbe2652aa7948242a57107ca7ab12fd2f8/llvm/lib/Target/RISCV/RISCVISelLowering.cpp#L5290-L5297

Reply:
@wangpc-pp BPI-F3 there for vrgather at `e8m1` can process eight elements in a cycle, and four at `e16m1`. Milk-V Pioneer at `e8m1` can do 32 elements per cycle, `e16m1` - 16 elts/cycle, `e32m1` - 8 elts/cycle, and even at `e64m1` four elements per cycle (yes, that's XTheadVector not RVV1.0, but IIRC an equivalent RVV1.0 architecture is supposed to come out at some point). These all should easily exceed the throughput of indexed loads on these architectures, even though the data on those aren't present. (things indeed get less pretty at higher LMUL though; I suppose cost models would be useful here)

Theoretically future architectures could check the index list of indexed loads/stores and special-case patterns, but I haven't heard of any RVV implementations doing such, and, seeing as x86-64 has had indexed loads for a decade (and SVE to a smaller extent) but nothing like this optimization, I wouldn't bet on that becoming widespread. And even if it was optimized, it would have to be doing something equivalent to `vrgather` internally anyway.

Reply:
> These cases should easily exceed the throughput of indexed loads on these architectures

vrgather.vv doesn't perform the load though so I'm not sure if we can compare them directly. The vluxei*.v is kinda doing two in one. 

I think the big performance concern is LMUL > 1, according to https://camel-cdr.github.io/rvv-bench-results/bpi_f3/index.html it's 16 cycles at e8m2 and 64 at e8m4 on the BPI-F3. The loop vectorizer uses LMUL 2 by default, if it ever learns to emit expanded loads.

Reply:
Thanks for the discussion! I think we may leave a TODO here to investigate which way is better:
1. `vcpop.m+vleN.v+viota.m+vrgather.vv`
```asm
vsetivli zero, 16, e8, m1, ta, ma
vcpop.m a1, v0
vsetvli zero, a1, e32, m4, ta, ma
vle32.v v12, (a0)
vsetivli zero, 16, e32, m4, ta, mu
viota.m v16, v0
vrgather.vv v8, v12, v16, v0.t
ret
```
It has:
* Consecutive unit-stride load, which is fast.
* Larger register pressure. We need to retain v12+v16 for above example.
* More vtype toggles (2).
* vrgather.vv may be slow for large LMUL.
2. `viota.m+vsll.vi(optional)+vluxeiN.v`
```asm
vsetivli zero, 16, e32, m4, ta, ma
viota.m v12, v0
vsll.vi v12, v12, 2, v0.t
vsetvli zero, zero, e32, m4, ta, mu
vluxei32.v v8, (a0), v12, v0.t
ret
```
It has:
* Indexed load, which is slower.
* Small register pressure. We only use v12 for above example.
* Fewer instructions (2 or 3) and vtype toggles (only 1).

Currently I prefer the second, but I don't have performance data to support this.

Reply:
Yeah, I didn't add the index load/store to byteswap, because I saw it perform badly in the LUT4 benchmark, and you can already infer the performance from that.

I think the optimal implementation for a reverse would be LMUL=8 (or 4 if we want to be conservative) uni-stride load/store + 8/4 times unrolled LMUL=1 vrgather.vv.
The byteswap benchmark is basically the same case.

You can express a LMUL>1 vrgather operation that doesn't need to cross lanes, in multiple LMUL=1 vrgather operations, which gives you a O(N) scale with LMUL for vrgather.
I've only seen a single case where indexed loads beat vrgather, it's not by much (20%), and only once we are memory bound, otherwise vrgather is much faster (up to 4x): https://camel-cdr.github.io/rvv-bench-results/bpi_f3/LUT4.html
I presume this is caused by the particular processor supporting chaining for indexed loads, but not for vrgather.

Reply:
Running some experiments on the Banana Pi F3, vrgather.vv is indeed about 4x faster at LMUL 1 and still about 2x faster at LMUL 4. At LMUL 8 the difference seems to level off.

Given this is going to be specific to the microarchitecture maybe we should have a subtarget feature to use vrgather.vv here, enabled by default for the spacemit-k1. I would like to be optimistic that at some point there will be hardware with a faster vluxei*.v implementation.